cmd/snap-update-ns/secure_bindmount.go

```diff
@@ -59,36 +54,44 @@ func (sec *Secure) BindMount(sourceDir, targetDir string, flags uint, stashDir s
 	}
 	defer sysClose(targetFd)
 
-	// Step 2: chdir to the source, and bind mount "." to the stash dir
+	// Step 2: perform a bind mount between the paths identified by the two
+	// file descriptors. We primarily care about privilege escalation here and
+	// trying to race the sysMount() by removing any part of the dir (sourceDir
+	// or targetDir) after we have an open file descriptor to it (sourceFd or
+	// targetFd) to then replace an element of the dir's path with a symlink
+	// will cause the fd path (ie, sourceFdPath or targetFdPath) to be marked
+	// as unmountable within the kernel (this path is also changed to show as
+	// '(deleted)'). Alternatively, simply renaming the dir (sourceDir or
+	// targetDir) after we have an open file descriptor to it (sourceFd or
+	// targetFd) causes the mount to happen with the newly renamed path, but
+	// this rename is controlled by DAC so while the user could race the mount
+	// source or target, this rename can't be used to gain privileged access to
+	// files. For systems with AppArmor enabled, this raced rename would be
+	// denied by the per-snap snap-update-ns AppArmor profle.
+	sourceFdPath := fmt.Sprintf("/proc/self/fd/%d", sourceFd)
+	targetFdPath := fmt.Sprintf("/proc/self/fd/%d", targetFd)
 	bindFlags := syscall.MS_BIND | (flags & syscall.MS_REC)
-	if err := sysFchdir(sourceFd); err != nil {
-		return err
-	}
-	if err := sysMount(".", stashDir, "", uintptr(bindFlags), ""); err != nil {
+	if err := sysMount(sourceFdPath, targetFdPath, "", uintptr(bindFlags), ""); err != nil {
 		return err
 	}
-	defer sysUnmount(stashDir, syscall.MNT_DETACH|umountNoFollow)
 
 	// Step 3: optionally change to readonly
 	if flags&syscall.MS_RDONLY != 0 {
-		remountFlags := syscall.MS_REMOUNT | syscall.MS_BIND | syscall.MS_RDONLY
-		if flags&syscall.MS_REC != 0 {
-			remountFlags |= syscall.MS_REC
+		// We need to look up the target directory a second time, because
+		// targetFd refers to the path shadowed by the mount point.
+		mountFd, err := sec.OpenPath(targetDir)
+		if err != nil {
+			// FIXME: the mount occurred, but the user moved the target
+			// somewhere
+			return err
```

Comment:
I think this could be a fatal error, we have a struct for those. Then snap-update-ns will just say "okay, I'm out" quickly and we will not save the namespace.

Reply:
I'd guess all errors here are implicitly fatal, but the actual `FatalError` will be returned by the caller if `BindMount` fails. I think we already have this pattern in `execWritableMimic`.